tslearn/nonmyopic.py

```diff
@@ -0,0 +1,370 @@
+            self.clusters_["ck_cm{0}".format(k)] = []
+            self.pyhatyck_["pyhatycks{0}".format(k)] = []
+            self.indice_ck_.append(np.where(c_k == k))
+            current_sum = self.pyck_[:, k].sum()
+            for current_classe in range(0, self.__n_classes_):
+                self.pyck_[current_classe, k] = self.pyck_[current_classe, k] / current_sum
+
+        for t in range(self.minimum_time_stamp, X.shape[1] + 1):
+            self.classifier_.append(
+                self.classifier(
+                    solver=self.solver,
+                    hidden_layer_sizes=self.hidden_layer_sizes,
+                    random_state=self.random_state,
+                    max_iter=self.maximum_iteration,
+                )
+            )
+            self.classifier_[-1].fit(X1[:, :t], Y1)
+            for k in range(0, self.n_clusters):
+                index_cluster = np.where(c_k2 == k)
+                if len(index_cluster[0]) != 0:
+                    X2_current_cluster = np.squeeze(X2[index_cluster, :t], axis=0)
+                    Y2_current_cluster = Y2[tuple(index_cluster)]
+                    Y2_hat = self.classifier_[-1].predict(X2_current_cluster[:, :t])
+                    self.clusters_["ck_cm{0}".format(k)].append(
+                        confusion_matrix(Y2_current_cluster, Y2_hat, labels=classes_y)
+                    )
+                    column_sum = self.clusters_["ck_cm{0}".format(k)][
+                        -1
+                    ].sum(1)
+                    for current_class in range(0, self.__n_classes_):
+                        if column_sum[current_class] == 0:
+                            column_sum[current_class] = 1
+                    matrix_proba_current_cluster_class = np.asarray(
+                        [
+                            [
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    0
+                                ][0]
+                                / column_sum[0],
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    0
+                                ][1]
+                                / column_sum[0],
+                            ],
+                            [
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    1
+                                ][0]
+                                / column_sum[1],
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    1
+                                ][1]
+                                / column_sum[1],
+                            ],
+                        ]
+                    )
+                    self.pyhatyck_["pyhatycks{0}".format(k)].append(
+                        matrix_proba_current_cluster_class
+                    )
+
+    def get_cluster_probas(self, X):
```

Comment:
This function could definitely be vectorized. I guess we need no `for` loop nor lists in here.

For distance computations:

```python
diffs = X[numpy.newaxis, :] - centroids
distances_clusters = numpy.linalg.norm(diffs, axis=(1, 2))
```

Then, similarly, `Delta_k` and `probas` can be vectorized:

```python
delta = (average_distance - distances_clusters) / average_distance
probas = 1. / (np.exp(self.lamb * minimum_distance_to_average)
               + np.exp(self.lamb * (minimum_distance_to_average - delta))
               + 1e-6)
return probas / probas.sum()
```